presto-main/src/main/java/com/facebook/presto/type/DoubleOperators.java

```diff
@@ -238,6 +239,28 @@ public static long hashCode(@SqlType(StandardTypes.DOUBLE) double value)
     }
 
     @ScalarOperator(SATURATED_FLOOR_CAST)
+    @SqlType(StandardTypes.FLOAT)
+    public static long saturatedFloorCastToFloat(@SqlType(StandardTypes.DOUBLE) double value)
+    {
+        float result;
+        float minFloat = -1.0f * Float.MAX_VALUE;
+        if (value <= minFloat) {
+            result = minFloat;
+        }
+        else if (value >= Float.MAX_VALUE) {
+            result = Float.MAX_VALUE;
+        }
+        else {
+            result = (float) value;
```

Comment:
This logic seems fragile, and I'm not even sure it's correct for non-strict-fp math.

Take a look at the following sections of the Java Language spec:

* https://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.2.3

> Every implementation of the Java programming language is required to support two standard sets of floating-point values, called the float value set and the double value set. In addition, an implementation of the Java programming language may support either or both of two extended-exponent floating-point value sets, called the float-extended-exponent value set and the double-extended-exponent value set. These extended-exponent value sets may, under certain circumstances, be used instead of the standard value sets to represent the values of expressions of type float or double

* https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.4:

> Within an expression that is not FP-strict, some leeway is granted for an implementation to use an extended exponent range to represent intermediate results